Make PhysicalFilesWatcher exclusively use polling when pollForChanges=true

ericstj opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/37664

ericstj committed:
Test out INotify exhaustion

ericstj committed:
Use DisposableFileSystem and update wait time

maryamariyan commented:
The PR is giving a repro of the error.

https://dev.azure.com/dnceng/public/_build/results?buildId=789722&view=ms.vss-test-web.build-test-results-tab&runId=24915252&resultId=144507&paneView=debug

```
System.IO.IOException : The configured user limit (1024) on the number of inotify instances has been reached, or the per-process limit on the number of open file descriptors has been reached.
```

Stack trace
```
at System.IO.FileSystemWatcher.StartRaisingEvents() in /_/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Linux.cs:line 46 at System.IO.FileSystemWatcher.StartRaisingEventsIfNotDisposed() in /_/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs:line 638 at System.IO.FileSystemWatcher.set_EnableRaisingEvents(Boolean value) in /_/src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs:line 151 at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.TryEnableFileSystemWatcher() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:line 377 at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.CreateFileChangeToken(String filter) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:line 133 at Microsoft.Extensions.FileProviders.PhysicalFileProvider.Watch(String filter) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFileProvider.cs:line 335 at Microsoft.Extensions.FileProviders.PhysicalFileProviderTests.PollingFileProviderShouldntConsumeINotifyInstances() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:line 113 at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__140_0(Object state) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs:line 1883
```

cc: @pranavkm

maryamariyan committed:
Fixing inotify instance limit exhaustion

maryamariyan commented:
@pranavkm do you see any possible risk with making this fix? https://github.com/dotnet/runtime/commit/e8669aab0e5a1dc2e960077326c757e060e26de3

pranavkm commented:
> @pranavkm do you see any possible risk with making this fix? e8669aa

That looks great 👍

ericstj commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -87,6 +89,50 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
+            int instances = maxInstances + 16;
+            try
+            {
+                // right now this will fail since we still use FSW
```
We should remove this comment

ericstj commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -87,6 +89,50 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
         }
 
+        [Fact]
+        [PlatformSpecific(TestPlatforms.AnyUnix)]
```
```suggestion
        [PlatformSpecific(TestPlatforms.Linux)] 
```

ericstj commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -87,6 +89,50 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
+                }
+            }
+        }
+
```
```suggestion
```

ericstj committed:
Respond to feedback

ericstj commented:
@pranavkm how likely do you imagine folks are to be regressed by this change?  Since we aren't enabling the watcher it means that the only events will be raised by polling.  This could result in latency or different events entirely if polling doesn't capture all types of changes that a FileSystemWatcher would.  What do you think about the likelihood of this?

ericstj commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:
```diff
@@ -130,7 +130,10 @@ public IChangeToken CreateFileChangeToken(string filter)
 
             IChangeToken changeToken = GetOrAddChangeToken(filter);
-            TryEnableFileSystemWatcher();
+            if (!PollForChanges)
```
It might be clearer if we just made the constructor ignore the FileSystemWatcher entirely.  Moreover, I wonder if we should be adding a constructor overload which doesn't require a `FileSystemWatcher`.

ericstj replied:
Per offline conversation to make this less risky we'll make the constructor allow for null FSW.  In the case of a null FSW and PollForChanges=true we'll permit only polling mode.  We'll make PFP construct the type this way when UsePolling is true.

pranavkm commented:
@ericstj we've [advertised the flag that enables polling](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/file-providers?view=aspnetcore-3.1#watch-for-changes) as a resort for apps that are deployed to environments that do not trigger the .NET file watcher (e.g. mounted drives, network shares etc). For such users, this change would make no visible difference.

eerhardt commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -87,6 +89,48 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
+                root.CreateFile("test.txt");
+
+                // let some events fire
+                await Task.Delay(WaitTimeForTokenToFire * 16);
```
Is this an 8 second delay? Is that a little too long?

ericstj replied:
Previously only had 500ms and it wasn't enough.  I'll switch to an event so that in the passing scenario it doesn't wait too long.

eerhardt commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -87,6 +89,48 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
+                // let some events fire
+                await Task.Delay(WaitTimeForTokenToFire * 16);
+
+                Assert.NotEqual(0, callbacks);
```
Instead of using a `int callbacks` count, would it make more sense to use an `AutoResetEvent` (or similar) that gets set in the callback?

ericstj replied:
yep, good suggestion

carlossanlop commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -87,6 +89,48 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
+            using var root = new DisposableFileSystem();
+
+            int callbacks = 0;
+            int maxInstances = int.Parse(File.ReadAllText("/proc/sys/fs/inotify/max_user_instances"));
```
Is it safe to assume that this file exists in this path for all distros?

carlossanlop commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -87,6 +89,48 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
+            using var root = new DisposableFileSystem();
+
+            int callbacks = 0;
+            int maxInstances = int.Parse(File.ReadAllText("/proc/sys/fs/inotify/max_user_instances"));
```
Is it safe to assume that this file exists in this path for all distros?

carlossanlop replied:
Oh, I just noticed that the test was originally topped with `TestPlatforms.AnyUnix`. Since `Linux` is a more specific platform, I assume we can confidently expect the file to exist in the test environment for that platform.

ericstj replied:
I still went ahead and added an assertion for the file existence to ensure we fail gracefully when absent.

ericstj committed:
Make PhysicalFileProvider decide if FSW will be used

This is less breaking that ignoring the FileSystemWatcher when polling
is set as it still lets the PhysicalFileWatcher work in FSW + Polling
mode.

Now when a FileSystemWatcher is not passed to the PhysicalFileWatcher
we'll permit construction (if polling is enabled) and have it behave as
exclusively polling.

The PhysicalFileProvider will use this mode when both
UsePollingFileWatcher and UseActivePolling are set which is what happens
when the DOTNET_USE_POLLING_FILE_WATCHER environment variable is set.

eerhardt commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:
```diff
@@ -79,17 +79,26 @@ public class PhysicalFilesWatcher : IDisposable
             PollForChanges = pollForChanges;
             _filters = filters;
 
+            if (fileSystemWatcher == null && !pollForChanges)
```
(nit) why do this arg checking in the middle of the ctor? Why not do it at the top of the method?

ericstj replied:
I guess I felt it was more of an incompatible check rather than a guard.  I suppose if this type did any argument checking I would have put it with that, so I can move.  What's your thought on the use of `ArgumentNullException` for this case?  Previously it was a NullRef.

eerhardt replied:
I think the `ArgumentNullException` makes sense. If this `bool` is `false`, then this arg can't be `null`...

eerhardt reviewed (approved):
:shipit:

stephentoub reviewed (approved):
Other than the existing comment on guard check placement, LGTM.

ericstj committed:
Address PR feedback

ericstj committed:
Merge remote-tracking branch 'upstream/master' into pfp.polling

# Conflicts:
#	src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs

ericstj committed:
Tweak test threshold

ericstj commented:
Test failures as follows:
```
Unhandled Exception:
System.IO.DirectoryNotFoundException: Could not find a part of the path '/tmp/aexy3v3e.mis'.
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].CreateDirectoryHandle(String path, Boolean ignoreNotFound) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs:line 79
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].Init() in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs:line 39
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]..ctor(String directory, Boolean isNormalized, EnumerationOptions options) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.cs:line 41
   at System.IO.Enumeration.FileSystemEnumerable`1.DelegateEnumerator[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]..ctor(FileSystemEnumerable`1 enumerable, Boolean isNormalized) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerable.cs:line 66
   at System.IO.Enumeration.FileSystemEnumerable`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerable.cs:line 38
   at System.IO.Enumeration.FileSystemEnumerableFactory.FileSystemInfos(String directory, String expression, EnumerationOptions options, Boolean isNormalized) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerableFactory.cs:line 203
   at System.IO.DirectoryInfo.InternalEnumerateInfos(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options) in /_/src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs:line 190
   at System.IO.DirectoryInfo.EnumerateFileSystemInfos(String searchPattern, EnumerationOptions enumerationOptions) in /_/src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs:line 172
   at System.IO.DirectoryInfo.EnumerateFileSystemInfos(String searchPattern, SearchOption searchOption) in /_/src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs:line 169
   at Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoWrapper.EnumerateFileSystemInfos()+MoveNext() in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Abstractions/DirectoryInfoWrapper.cs:line 38
   at System.Collections.Generic.List`1[[Microsoft.Extensions.FileSystemGlobbing.Abstractions.FileSystemInfoBase, Microsoft.Extensions.FileSystemGlobbing, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].InsertRange(Int32 index, IEnumerable`1 collection) in /_/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:line 742
   at System.Collections.Generic.List`1[[Microsoft.Extensions.FileSystemGlobbing.Abstractions.FileSystemInfoBase, Microsoft.Extensions.FileSystemGlobbing, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].AddRange(IEnumerable`1 collection) in /_/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:line 243
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Match(DirectoryInfoBase directory, String parentRelativePath) in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Internal/MatcherContext.cs:line 67
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Execute() in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Internal/MatcherContext.cs:line 53
   at Microsoft.Extensions.FileSystemGlobbing.Matcher.Execute(DirectoryInfoBase directoryInfo) in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Matcher.cs:line 166
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.CalculateChanges() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingWildCardChangeToken.cs:line 108
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.get_HasChanged() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingWildCardChangeToken.cs:line 98
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:line 437
   at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 627
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 589
   at System.Threading.TimerQueueTimer.System.Threading.IThreadPoolWorkItem.Execute() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 573
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:line 646
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs:line 41
   at System.Threading.Thread.StartCallback() in /_/src/mono/netcore/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:line 278
[ERROR] FATAL UNHANDLED EXCEPTION: System.IO.DirectoryNotFoundException: Could not find a part of the path '/tmp/aexy3v3e.mis'.
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].CreateDirectoryHandle(String path, Boolean ignoreNotFound) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs:line 79
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]].Init() in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.Unix.cs:line 39
   at System.IO.Enumeration.FileSystemEnumerator`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]..ctor(String directory, Boolean isNormalized, EnumerationOptions options) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerator.cs:line 41
   at System.IO.Enumeration.FileSystemEnumerable`1.DelegateEnumerator[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]..ctor(FileSystemEnumerable`1 enumerable, Boolean isNormalized) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerable.cs:line 66
   at System.IO.Enumeration.FileSystemEnumerable`1[[System.IO.FileSystemInfo, System.IO.FileSystem, Version=5.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a]]..ctor(String directory, FindTransform transform, EnumerationOptions options, Boolean isNormalized) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerable.cs:line 38
   at System.IO.Enumeration.FileSystemEnumerableFactory.FileSystemInfos(String directory, String expression, EnumerationOptions options, Boolean isNormalized) in /_/src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEnumerableFactory.cs:line 203
   at System.IO.DirectoryInfo.InternalEnumerateInfos(String path, String searchPattern, SearchTarget searchTarget, EnumerationOptions options) in /_/src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs:line 190
   at System.IO.DirectoryInfo.EnumerateFileSystemInfos(String searchPattern, EnumerationOptions enumerationOptions) in /_/src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs:line 172
   at System.IO.DirectoryInfo.EnumerateFileSystemInfos(String searchPattern, SearchOption searchOption) in /_/src/libraries/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs:line 169
   at Microsoft.Extensions.FileSystemGlobbing.Abstractions.DirectoryInfoWrapper.EnumerateFileSystemInfos()+MoveNext() in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Abstractions/DirectoryInfoWrapper.cs:line 38
   at System.Collections.Generic.List`1[[Microsoft.Extensions.FileSystemGlobbing.Abstractions.FileSystemInfoBase, Microsoft.Extensions.FileSystemGlobbing, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].InsertRange(Int32 index, IEnumerable`1 collection) in /_/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:line 742
   at System.Collections.Generic.List`1[[Microsoft.Extensions.FileSystemGlobbing.Abstractions.FileSystemInfoBase, Microsoft.Extensions.FileSystemGlobbing, Version=5.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60]].AddRange(IEnumerable`1 collection) in /_/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs:line 243
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Match(DirectoryInfoBase directory, String parentRelativePath) in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Internal/MatcherContext.cs:line 67
   at Microsoft.Extensions.FileSystemGlobbing.Internal.MatcherContext.Execute() in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Internal/MatcherContext.cs:line 53
   at Microsoft.Extensions.FileSystemGlobbing.Matcher.Execute(DirectoryInfoBase directoryInfo) in /_/src/libraries/Microsoft.Extensions.FileSystemGlobbing/src/Matcher.cs:line 166
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.CalculateChanges() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingWildCardChangeToken.cs:line 108
   at Microsoft.Extensions.FileProviders.Physical.PollingWildCardChangeToken.get_HasChanged() in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PollingWildCardChangeToken.cs:line 98
   at Microsoft.Extensions.FileProviders.Physical.PhysicalFilesWatcher.RaiseChangeEvents(Object state) in /_/src/libraries/Microsoft.Extensions.FileProviders.Physical/src/PhysicalFilesWatcher.cs:line 437
   at System.Threading.TimerQueueTimer.CallCallback(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 627
   at System.Threading.TimerQueueTimer.Fire(Boolean isThreadPool) in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 589
   at System.Threading.TimerQueueTimer.System.Threading.IThreadPoolWorkItem.Execute() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/Timer.cs:line 573
   at System.Threading.ThreadPoolWorkQueue.Dispatch() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:line 646
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart() in /_/src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPool.WorkerThread.cs:line 41
   at System.Threading.Thread.StartCallback() in /_/src/mono/netcore/System.Private.CoreLib/src/System/Threading/Thread.Mono.cs:line 278
```

This indicates a problem with the polling file watcher if the directory it is watching is deleted underneath it.  This is a reliability issue and one we should probably address.  Without it the tests I added will be flaky.

ericstj committed:
Merge branch 'master' of https://github.com/dotnet/runtime into pfp.polling

ericstj committed:
Add test around deleting watched directory.

ericstj committed:
Make directory delete test use polling

We have to avoid exclusively using FileSystemWatcher since it looks like
FSW misses file deletes that happen as a result of directory delete on OSX.

ericstj commented:
Ok, the flakiness in this test should be fixed now.  Can I get another review?

eerhardt commented on src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs:
```diff
@@ -1479,5 +1530,31 @@ public void CreateFileWatcher_CreatesWatcherWithPollingAndActiveFlags()
+            using (var provider = new PhysicalFileProvider(root.RootPath))
+            {
+                var fileName = Path.GetRandomFileName();
+                var fileLocation = Path.Combine(root.RootPath, fileName);
```
I don't see a usage of `fileLocation`.

ericstj replied:
Good catch, I had that in a previous version, but had to remove it due to #44484.  Will clean it up.

eerhardt reviewed (approved):
I just had one tiny nit. LGTM.

ericstj committed:
Remove dead code from test.
